File: jvm/optimized_version.java
package jvm;

import java.util.Random;

public class optimized_version {

    public static void main(String[] args) {
        var u = Integer.parseInt(args[0]); // Get an input number from the command line
        var r = new Random().nextInt(10000); // Get a random number 0 <= r < 10k
        var a = new int[10000]; // Array of 10k elements initialized to 0
        var factor = 100000 / u;
        var remaining = 100000 % u;
        
        for (var i = 0; i < 10000; i++) { // 10k outer loop iterations
            for (var j = 0; j < factor; j++) { // 100k inner loop iterations, per outer loop iteration
                for (var k = 0; k < u; k++) {
                    a[i] = a[i] + k; // Simple sum
                }
            }

            for (var k = 0; k < remaining; k++) {
                a[i] = a[i] + k; // Simple sum
            }

            a[i] += r; // Add a random value to each element in array
        }
        System.out.println(a[r]); // Print out a single element from the array
    }
}

File: c/optimized_version.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <time.h>
#include <math.h>

int main (int argc, char** argv) {
  int u = atoi(argv[1]);               // Get an input number from the command line
  srand(time(NULL));                   // FIX random seed
  int r = rand() % 10000;              // Get a random integer 0 <= r < 10k
  int32_t a[10000] = {0};              // Array of 10k elements initialized to 0
  int factor = 100000 / u;
  int remaining = 100000 % u;

  for (int i = 0; i < 10000; i++) {    // 10k outer loop iterations
    for (int j = 0; j < factor; j++) { // 100k inner loop iterations, per outer loop iteration
      for (int k = 0; k < u; k++) {
        a[i] = a[i] + k;               // Simple sum
      }
    }

    for (int k = 0; k < remaining; k++) {
      a[i] = a[i] + k;
    }

    a[i] += r;                         // Add a random value to each element in array
  }
  printf("%d\n", a[r]);                // Print out a single element from the array
}